pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/generator/RealtimeToOfflineSegmentsTaskGenerator.java

```diff
@@ -0,0 +1,308 @@
+
+    for (TableConfig tableConfig : tableConfigs) {
+      String tableName = tableConfig.getTableName();
+
+      if (tableConfig.getTableType() != TableType.REALTIME) {
+        LOGGER.warn("Skip generating task: {} for non-REALTIME table: {}", taskType, tableName);
+        continue;
+      }
+      LOGGER.info("Start generating task configs for table: {} for task: {}", tableName, taskType);
+
+      String rawTableName = TableNameBuilder.extractRawTableName(tableName);
+      String realtimeTableName = TableNameBuilder.REALTIME.tableNameWithType(rawTableName);
+
+      // Only schedule 1 task of this type, per table
+      Map<String, TaskState> nonCompletedTasks =
+          TaskGeneratorUtils.getNonCompletedTasks(taskType, realtimeTableName, _clusterInfoProvider);
+      if (!nonCompletedTasks.isEmpty()) {
+        LOGGER.warn("Found non-completed tasks: {} for same table: {}. Skipping task generation.",
+            nonCompletedTasks.keySet(), realtimeTableName);
+        continue;
+      }
+
+      // Get all completed segment metadata.
+      List<LLCRealtimeSegmentZKMetadata> completedSegmentsMetadata = new ArrayList<>();
+      Map<Integer, String> partitionToLatestCompletedSegmentName = new HashMap<>();
+      Set<Integer> allPartitions = new HashSet<>();
+      getCompletedSegmentsInfo(realtimeTableName, completedSegmentsMetadata, partitionToLatestCompletedSegmentName,
+          allPartitions);
+      if (completedSegmentsMetadata.isEmpty()) {
+        LOGGER
+            .info("No realtime completed segments found for table: {}, skipping task generation: {}", realtimeTableName,
+                taskType);
+        continue;
+      }
+      allPartitions.removeAll(partitionToLatestCompletedSegmentName.keySet());
+      if (!allPartitions.isEmpty()) {
+        LOGGER
+            .info("Partitions: {} have no completed segments. Table: {} is not ready for {}. Skipping task generation.",
+                allPartitions, realtimeTableName, taskType);
+        continue;
+      }
+
+      TableTaskConfig tableTaskConfig = tableConfig.getTaskConfig();
+      Preconditions.checkState(tableTaskConfig != null);
+      Map<String, String> taskConfigs = tableTaskConfig.getConfigsForTaskType(taskType);
+      Preconditions.checkState(taskConfigs != null, "Task config shouldn't be null for table: {}", tableName);
+
+      // Get the bucket size and buffer
+      String bucketTimeStr =
+          taskConfigs.getOrDefault(RealtimeToOfflineSegmentsTask.BUCKET_TIME_PERIOD_KEY, DEFAULT_BUCKET_PERIOD);
+      String bufferTimeStr =
+          taskConfigs.getOrDefault(RealtimeToOfflineSegmentsTask.BUFFER_TIME_PERIOD_KEY, DEFAULT_BUFFER_PERIOD);
+      String startTimeStr = taskConfigs.get(RealtimeToOfflineSegmentsTask.START_TIME_MILLIS_KEY);
+      long bucketMillis = TimeUtils.convertPeriodToMillis(bucketTimeStr);
+      long bufferMillis = TimeUtils.convertPeriodToMillis(bufferTimeStr);
```

Comment:
Same for other `Millis`
```suggestion
      long bucketTimeMs = TimeUtils.convertPeriodToMillis(bucketTimeStr);
      long bufferTimeMs = TimeUtils.convertPeriodToMillis(bufferTimeStr);
```